Allow HTTP::Client to use ipv6 addresses

bcardiff opened the pull request:
Fixes #5514 

The `URI#host` is fine to have the brackets since they are part of rfc3986 host description.
The `HTTP::Client` needs to extract them before creating the `TCPSocket`.

j8r commented on src/http/client.cr:
```diff
@@ -651,7 +651,11 @@ class HTTP::Client
+    # URI host may have ipv6 enclosed by brackets and is fine according to
+    # rfc3986. The host http header must contains the brackets, ie: the same
+    # value as returned by URI#host
+    ip_host = @host[0] == '[' && @host[-1] == ']' ? @host[1..-2] : @host
```
`@host.starts_with?('[') && @host.ends_with?(']')` is x2 faster.
`@host.lstrip.rstrip` is said to be 7x than `@host[1..-2]`. I wonder if adding a method to strip both at the start and at the end can make sense.

```crystal
host = "[test]"

Benchmark.ips do |x|
  x.report("index") do
    host[1..-2]
  end

  x.report("strip") do
      host.lstrip.rstrip
  end                              
end

index  29.23M ( 34.21ns) (± 4.66%)  7.72× slower
     strip 225.75M (  4.43ns) (± 3.75%)       fastest
```

RX14 replied:
Thats because `lstrip` doesn't do what you want here?

RX14 replied:
`.lchop('[').rchop(']')` should be shorter here though. Surely the same speed.

j8r replied:
Sorry for the mistake, I confuse the two. Change `strip` to `chop` to have `lchop.rchop`
With chops the program is slower

RX14 commented on src/http/client.cr:
```diff
@@ -651,7 +651,11 @@ class HTTP::Client
+    # URI host may have ipv6 enclosed by brackets and is fine according to
+    # rfc3986. The host http header must contains the brackets, ie: the same
+    # value as returned by URI#host
+    ip_host = @host[0] == '[' && @host[-1] == ']' ? @host[1..-2] : @host
```
Both Ruby and Go expose a `hostname`, which is just the `host` without `[]` if it's ipv6. Perhaps we should copy them.

bcardiff commented:
I moved the unwrap to URI, but since in http client `@host` is `String`, it needs to be `URI.hostname(host : String)`. But I also added `URI#hostname` nevertheless.

Regarding benchmark, the original code wins:

```cr
require "benchmark"

host = "[test#{Random.new.next_int}]"
pp host
pp host[0] == '[' && host[-1] == ']' ? host[1..-2] : host

Benchmark.ips do |x|
  x.report("index") do
    host[0] == '[' && host[-1] == ']' ? host[1..-2] : host
  end

  x.report("chop") do
    host.lchop('[').rchop(']')
  end
end
```

```
$ crystal bench.cr --release
host # => "[test778306676]"
(host[0] == '[') && (host[-1] == ']') ? host[1..-2] : host # => "test778306676"
index  15.24M ( 65.62ns) (± 7.53%)       fastest
 chop  10.08M ( 99.24ns) (± 8.62%)  1.51× slower
```

j8r commented:
@bcardiff Sorry for the mistake. `@host.starts_with?('[') && @host.ends_with?(']')` seems faster, at least it is on my side.

RX14 commented:
We can remove the ugly `URI.hostname` if we move the normalization of `@host` to the `HTTP::Client` constructor which takes `URI`.

bcardiff commented:
@RX14 that won't work. There are initializers that use String and not URI.
But I am ok to mark `URI.hostname` as `:nodoc:`.

RX14 commented:
@bcardiff that's fine - if someone passes `HTTP::Client.new("[...]")` they can fail.

bcardiff commented:
@RX14 I don't agree. clients initialized with `HTTP::Client.new("[::1]")` will work. And actually HTTP::Client.new("::1") won't, or at least it is not guaranteed. The host http header are supposed to have a `[::1]` and not `::1`. So the `@host` needs to have the brackets as in URI.

Otherwise we need to add logic to wrap the host upon client. It can be done, but is a different issue. 

Bottom line `@host`in  HTTP::Client is expected to match `URI#host`.

RX14 commented:
Well I really hate `URI.hostname`, so let's just open-code the removal of the `[]` into the method, like this was previously.

Then add `URI#hostname` in this or another PR.

bcardiff commented:
But that would lead to two implementation of the same thing. Let's leave `URI.hostname` as `:nodoc:` and already have `URI#hostname`.

I'm satisfied with the state of the PR.

RX14 commented:
A bit of copy and paste really isn't that bad, especially for a snippet of code so small and unlikely to change as this.

ysbaddaden commented:
I agree. Copying once is okayish. Copying a second time should trigger a refactor, but once is acceptable.

bcardiff commented:
Done, the PR should be squashed on merge. (I have the feeling that URI.hostname will come back sooner that later, but ok)
